### PR TITLE
Add datasource picker to import saved object flyout when multiple data source is enabled

### PR DESCRIPTION
**Pull request by BionIT**
### Description

When multiple data source plugin is enabled, user can connect to remote data sources in addition to the local cluster, to support importing saved objects from remote data source via file upload(https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5712), we need to have an option in the flyout for user to select where the file comes from. 

Via discussion in the issue, we are altering the order of the import option, and have the `Create new objects with random IDs`  as the default choice. When multiple data source is enabled, we will have the data source picker under `Import options` section then the `Conflict management` section will contain the options to address conflicts.

### Issues Resolved

Fixes https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5712 

## Screenshot

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/17714150/1bdda995-47f0-425e-820c-ce031f80fbf8

## Testing the changes

In the video above, we tested the following scenarios
when multiple data source is disabled:
- upload same ndjson with `Create new objects with unique IDs` option is successful when there are duplicate objects exists
- upload same ndjson with `Automatically overwrite conflicts` option is successful when there are duplicate objects
- upload same ndjson with `Request action on conflict` option will prompt user to handle on the conflict
- upload same json with `Automatically overwrite conflicts` option twice is successful when there are duplicate objects exists
- upload same json with `Request action on conflict` option will prompt user to handle on the conflict
when multiple data source is enabled:
- upload ndjson with `Create new objects with unique IDs` option is successful 
- upload ndjson with `Automatically overwrite conflicts` option is successful when there are duplicate objects
- upload same ndjson with `Request action on conflict` option will prompt user to handle on the conflict
- upload same json with `Automatically overwrite conflicts` option twice is successful when there are duplicate objects exists
- upload same json with `Request action on conflict` option will prompt user to handle on the conflict

### Check List

- [x] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Review by ashwin-pc — CHANGES_REQUESTED** [review]
We already have a datasource selector component in the data plugin. We should use that and not create a new picker.

**Comment by bandinib-amzn** [comment]
I'm coming here after watching video attached in screenshot. Does it make sense to add a checkbox or another option to select 'overwrite all' instead of having to click 'overwrite' 15 times? cc: @kgcreative

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx [hunk 1]
```diff
@@ -78,8 +79,10 @@ import { FailedImportConflict, RetryDecision } from '../../../lib/resolve_import
 import { OverwriteModal } from './overwrite_modal';
 import { ImportModeControl, ImportMode } from './import_mode_control';
 import { ImportSummary } from './import_summary';
+// eslint-disable-next-line @osd/eslint/no-restricted-paths
```
Comment (bandinib-amzn) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx:82]: Do we need this?

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx [hunk 2]
```diff
@@ -78,8 +79,10 @@ import { FailedImportConflict, RetryDecision } from '../../../lib/resolve_import
...
+import { DataSourcePicker } from '../../../../../data_source_management/public/components/data_source_picker/data_source_picker.js';
```
Comment (bandinib-amzn) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx:83]: nit: consider adding index file in `data_source_management/public/components/data_source_picker` and export from there

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx [hunk 3]
```diff
@@ -78,8 +79,10 @@ import { FailedImportConflict, RetryDecision } from '../../../lib/resolve_import
...
 
-const CREATE_NEW_COPIES_DEFAULT = false;
+const CREATE_NEW_COPIES_DEFAULT = true;
```
Comment (bandinib-amzn) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx:85]: why are we changing the value of this const? won't it impact current behaviour?

**Comment by bandinib-amzn** [comment]
> We already have a datasource selector component in the data plugin. We should use that and not create a new picker.

Hi @ashwin-pc 
I guess you are talking about this PR - https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5167 and wondering why we don't use the same selector.

Let's consider the use case from above PR which is Datasource selector support in OpenSearch Dashboard Discover.
The **Discover** feature in OpenSearch Dashboards is designed for searching, exploring, and visualizing data stored in **OpenSearch indices** or **Remote cluster indices**. 

The **Dev Tool** feature in OpenSearch Dashboard provides an environment for developers and administrators to interact with **OpenSearch** or **Remote cluster** with query DSL and API directly.

I took discover and dev tool here as an example to show the difference between two pickers here. They serve different purposes. #5167 has  picker for selecting data source for searching, exploring data in the selected data source. Whereas data source picker in this PR is listing the data sources (basically remote cluster) to interact with them directly for fetch sample data, fetch indices to create index patterns etc.

IMO, they both serves different purpose, so I would prefer separate picker. @kgcreative I would like to here your thoughts as well.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx (hunk 3 above)
Comment (BionIT) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx:85]: Yeah, as discussed with @kgcreative , we want to have create new copies with unique IDs as default option, and find screenshot here https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5712#issuecomment-1922650607, and this option is switched up in order as well

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx (hunk 1 above)
Comment (BionIT) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx:82]: Yes, this is required to avoid lint error

**Comment by BionIT** [comment]
> I'm coming here after watching video attached in screenshot. Would it be possible to consider adding a checkbox or an alternative option for selecting 'overwrite all,' instead of having to click 'overwrite' 15 times? This could potentially enhance the user experience. cc: @kgcreative

**Comment by BionIT** [comment]
> I'm coming here after watching video attached in screenshot. Would it be possible to consider adding a checkbox or an alternative option for selecting 'overwrite all,' instead of having to click 'overwrite' 15 times? This could potentially enhance the user experience. cc: @kgcreative

@bandinib-amzn The reason user has to act on the action multiple times is because `Request action on conflict` is selected.  If user selects `Automatically overwrite conflicts` then all the conflict objects will be overwritten without any additional actions. This is the current behavior of the import function

**Comment by kgcreative** [comment]
> I'm coming here after watching video attached in screenshot. Would it be possible to consider adding a checkbox or an alternative option for selecting 'overwrite all,' instead of having to click 'overwrite' 15 times? This could potentially enhance the user experience. cc: @kgcreative

Yes, this is ~PAINFUL~ today, however, i'm hesitant to add to the scope unless the dev effort is reasonable. Ideally, we should show a selectable list, with easy "Select all/Deselect all/multiselect" options. Presumably each modal in the current experience triggers only when there is a conflict, so this would require a re-think of the flow.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx (hunk 2 above)
Comment (ashwin-pc) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx:83]: +1

**Comment by BionIT** [comment]
> We already have a datasource selector component in the data plugin. We should use that and not create a new picker.

As discussed with @ashwin-pc and @mengweieric , we will switch to use the data source selector in 2.13 when it supports the feature requested in https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5790

**Comment by ashwin-pc** [comment]
> IMO, they both serves different purpose, so I would prefer separate picker. @kgcreative I would like to here your thoughts as well.

Thanks for that example. That was a very useful distinction! Then i think my concern is around the confusion in naming. We are calling this and the other component both datasource pickers but have very different uses for each. This one is used specifically to select an OpenSearch cluster while the other is to select what is in thes clusters but also data from other sources. Can we rename this to be a `cluster selector` instead to disambiguate the two?

**Comment by BionIT** [comment]
> > IMO, they both serves different purpose, so I would prefer separate picker. @kgcreative I would like to here your thoughts as well.
> 
> Thanks for that example. That was a very useful distinction! Then i think my concern is around the confusion in naming. We are calling this and the other component both datasource pickers but have very different uses for each. This one is used specifically to select an OpenSearch cluster while the other is to select what is in thes clusters but also data from other sources. Can we rename this to be a `cluster selector` instead to disambiguate the two?

@ashwin-pc I hear you, I don't have a strong opinion, since I will change to use the new data source selector in 2.13, should we go with this name for now? Happy to change it if we feel necessary

**Comment by ashwin-pc** [comment]
> @ashwin-pc I hear you, I don't have a strong opinion, since I will change to use the new data source selector in 2.13, should we go with this name for now? Happy to change it if we feel necessary

based on Bandini's reply i'm actually convinced that the two might not need to be the same thing if the only thing you are selecting is the cluster and not the index/indexPattern. But if thats true then we do need to rename this since the two wont be combined in 2.13 or beyond

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx (hunk 1 above)
Comment (bandinib-amzn) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx:82]: I think if we take care of [below comment](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5781#discussion_r1476707363), then we won't have lint errors.

**Comment by bandinib-amzn** [comment]
> e import function

Yes, I completely agree. It doesn't need to be within the scope of this PR. It's an enhancement that we can address in the future. I'll create an issue to track it.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx (hunk 3 above)
Comment (Flyingliuhub) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx:85]: I'm wondering if there any tests related to this in the [function test](https://github.com/opensearch-project/opensearch-dashboards-functional-test) repo if we changed the default behavior

Comment (Flyingliuhub) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx:85]: I'm wondering if there any tests related to this in the [function test ](https://github.com/opensearch-project/opensearch-dashboards-functional-test)repo if we changed the default behavior

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx (hunk 1 above)
Comment (BionIT) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx]: Moved export to top level and this line is removed in new commit

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx (hunk 2 above)
Comment (BionIT) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx]: Changed component name to cluster selector and also moved export to top level in new commit

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx (hunk 3 above)
Comment (BionIT) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/flyout.tsx:83]: The FT repo doesn't have tests for import saved objects, and the dashboard repo has functional tests which have been fixed to adopt the change of default behavior

**Comment by BionIT** [comment]
> Can we also add some tests for the code added here?

Added tests in new commit

**Review by Flyingliuhub — APPROVED** [review]

**Pull request merged**